transport/compress.go

```diff
@@ -0,0 +1,232 @@
+		if r.err != nil {
+			return 0, r.err
+		}
+
+		if r.err = readFull(r.r, r.buf[:r.l]); r.err != nil {
+			return 0, r.err
+		}
+
+		// TODO: Why we need to do this?
+		if zeroes(r.buf[:r.l]) {
+			return 0, errComprEnd
+		}
+
+		if r.err = r.decode(); r.err != nil {
+			return 0, r.err
+		}
+	}
+}
+
+func (r *comprReader) decodeSnappy() error {
+	n, err := s2.DecodedLen(r.buf)
+	if r.err != nil {
+		return err
+	}
+
+	if n > len(r.decoded) {
+		if n > r.maxBlock {
+			return ErrComprUnsupported
+		}
+		r.decoded = make([]byte, n)
+	}
+
+	buf := r.buf[:r.l]
+	if _, err := s2.Decode(r.decoded, buf); err != nil {
+		return err
+	}
+	r.i, r.j = 0, n
+
+	return nil
+}
+
+func (r *comprReader) decodeLz4() error {
+	var err error
+	n := binary.BigEndian.Uint32(r.buf[0:4])
+	r.decoded, err = ensureBufferSize(int(n), r.decoded, r.lazyBuf)
+	if r.err != nil {
+		return err
+	}
+
+	buf := r.buf[4:r.l]
+	if _, err := lz4.UncompressBlock(buf, r.decoded); err != nil {
+		return err
+	}
+	r.i, r.j = 0, int(n)
+	return nil
+}
+
+func ensureBufferSize(n int, buf []byte, maxSize int) ([]byte, error) {
+	if len(buf) >= n {
+		return buf, nil
+	}
+	if n > maxSize {
+		return nil, ErrComprCorrupt
+	}
+	// Realloc buffer.
+	buf = make([]byte, n)
+	return buf, nil
+}
+
+func readFull(r io.Reader, p []byte) error {
+	_, err := io.ReadFull(r, p)
+	return err
+}
+
+func zeroes(p []byte) bool {
+	for _, v := range p {
+		if v != 0 {
+			return false
+		}
+	}
+	return true
+}
+
+type comprWriter struct {
+	w          io.Writer
+	encoded    []byte
+	maxBufSize int
+	err        error
+	encode     func(p []byte) ([]byte, error)
+	// Only used in lz4 version:
+	c lz4.Compressor
+}
+
+func NewComprWriter(w io.Writer, isSnappy bool) *comprWriter {
+	cw := comprWriter{
+		w:          w,
+		maxBufSize: ioBufferSize,
+	}
+
+	switch isSnappy {
+	case true:
+		cw.encode = cw.encodeSnappy
+	default:
+		cw.encode = cw.encodeLz4
+	}
+	return &cw
+}
+
+func (w *comprWriter) Write(p []byte) (int, error) {
+	if len(p) == 9 {
+		return w.w.Write(p)
+	}
+
+	var out []byte
+	if out, w.err = w.encode(p); w.err != nil {
+		return 0, w.err
+	}
+
+	out[1] |= frame.Compress
+	return w.w.Write(out)
+}
+
+func (w *comprWriter) encodeSnappy(p []byte) (enc []byte, err error) {
+	dec := s2.Encode(w.encoded, p[9:])
+	if dec != nil {
+		if len(dec) > w.maxBufSize {
+			return nil, ErrComprUnsupported
+		}
+		w.encoded = dec
+	}
+
+	enc = p[:9]
+	enc = append(enc, w.encoded...)
+	binary.BigEndian.PutUint32(enc[5:9], uint32(len(w.encoded)))
+	return enc, nil
+}
+
+func (w *comprWriter) encodeLz4(p []byte) (enc []byte, err error) {
+	w.encoded, err = ensureBufferSize(lz4.CompressBlockBound(len(p)), w.encoded, w.maxBufSize)
+	if err != nil {
+		return nil, err
+	}
+
+	n, err := w.c.CompressBlock(p[9:], w.encoded)
+	if n == 0 || err != nil {
+		return nil, ErrComprUnsupported
+	}
+
+	l := make([]byte, 4, 4)
+	binary.BigEndian.PutUint32(l, uint32(len(p)))
+
+	enc = p[:9]
+	enc = append(enc, l...)
```

Comment:
Would be better to append shifted bytes directly, we do that in other places already.